Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDataRecordCocoa.mm

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#if !__has_feature(objc_arc)
+#error This file requires ARC. Add the "-fobjc-arc" compiler flag for this file.
+#endif
+
+#import "config.h"
+#import "WebExtensionDataRecord.h"
+
+#if ENABLE(WK_WEB_EXTENSIONS)
+
+#import "_WKWebExtensionDataRecordInternal.h"
+
+namespace WebKit {
+
+void WebExtensionDataRecord::setError(NSString *underlyingErrorMessage, WebExtensionDataType type)
+{
+    switch (type) {
+    case WebExtensionDataType::Local:
+        wrapper().localStorageError = createDataRecordError(_WKWebExtensionDataRecordErrorLocalStorageFailed, underlyingErrorMessage);
+        break;
+    case WebExtensionDataType::Session:
+        wrapper().sessionStorageError = createDataRecordError(_WKWebExtensionDataRecordErrorSessionStorageFailed, underlyingErrorMessage);
+        break;
+    case WebExtensionDataType::Sync:
+        wrapper().syncStorageError = createDataRecordError(_WKWebExtensionDataRecordErrorSyncStorageFailed, underlyingErrorMessage);
+        break;
```

Comment:
This seems backwards. The ObjC object should get the error from the C++ object. The C++ object should not be setting the object on the ObjC wrapper.

In this case `WebExtensionDataRecord` should have an `NSArray` of errors, similar to `WebExtension::errors()`.

Reply:
spoke about this offline with @xeenon. it's best to have WebExtensionDataRecord mirror _WKWebExtensionDataRecord. ideally, for all the web extension classes we've written, the c++ object should be a source of truth. with this approach, the _WKWebExtensionDataRecord would have these errors but the WebExtensionDataRecord wouldn't, which is not the behavior we want to promote.